### PR TITLE
ZGrid: add homing options

### PR DESCRIPTION
**Pull request by qharley**
Added homing options to ZGrid:
NOHOME
HOMEXY
HOMEXYZ (Default)

### CODE REVIEW

File: src/modules/tools/zprobe/ZGridStrategy.cpp [hunk 1]
```diff
@@ -534,8 +543,22 @@ void ZGridStrategy::normalize_grid()
...
 void ZGridStrategy::homexyz()
 {
-    Gcode gc("G28", &(StreamOutput::NullStream));
-    THEKERNEL->call_event(ON_GCODE_RECEIVED, &gc);
+
+  switch(this->home_before_probe) {
+    case NOHOME : return;
+
+    case HOMEXY : {
+                      Gcode gc("G28 X0 Y0", &(StreamOutput::NullStream));
```
Comment (sorki) [inline src/modules/tools/zprobe/ZGridStrategy.cpp]: This should be just "G28 X Y"

File: src/modules/tools/zprobe/ZGridStrategy.cpp [hunk 2]
```diff
@@ -447,8 +456,8 @@ void ZGridStrategy::setZoffset(float zval)
 
 bool ZGridStrategy::doProbing(StreamOutput *stream)  // probed calibration
 {
-    // home first
-    this->homexyz();
+    // home first using selected mode: NOHOME, HOMEXY, HOMEXYZ
+        this->homexyz();
```
Comment (sorki) [inline src/modules/tools/zprobe/ZGridStrategy.cpp]: whitespace error

**Comment by sorki** [comment]
Please use more descriptive commit message prefixed with subsystem - e.g.:

    zgrid: add homing options

    The machine can be told to home in one of the following modes:
    
        leveling-strategy.ZGrid-leveling.home_before_probe  2;   
        # 0: NOHOME  1: HOMEXY  2: HOMEXYZ (Default HOMEXYZ)

**Comment by qharley** [comment]
Happy... ;-)

**Comment by wolfmanjm** [comment]
No It is G28 X0 Y0. G28 X Y is illegal Gcode. We are trying to keep oiut GCode to spec

File: src/modules/tools/zprobe/ZGridStrategy.cpp (hunk 1 above)
Comment (wolfmanjm) [inline src/modules/tools/zprobe/ZGridStrategy.cpp]: no you are wrong it is G28 X0 Y0

Comment (qharley) [inline src/modules/tools/zprobe/ZGridStrategy.cpp]: Ok... I'll fix it again!


  [image: photo]
*Quentin Harley*
Founder, Reprap Morgan 3D
 p:+27120041883 | e:quentin@reprapmorgan.com | w:www.reprapmorgan.com | a:4
Burger Ave, Lyttelton Manor, Centurion

File: src/modules/tools/zprobe/ZGridStrategy.cpp [hunk 3]
```diff
@@ -534,8 +543,22 @@ void ZGridStrategy::normalize_grid()
...
 void ZGridStrategy::homexyz()
 {
-    Gcode gc("G28", &(StreamOutput::NullStream));
-    THEKERNEL->call_event(ON_GCODE_RECEIVED, &gc);
+
+  switch(this->home_before_probe) {
+    case NOHOME : return;
+
+    case HOMEXY : {
+        Gcode gc("G28 X Y", &(StreamOutput::NullStream));
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/ZGridStrategy.cpp]: This was correct before it must be G28 X0 Y0. X Y is illegal gcode and we are trying to keep out GCode to spec. This is not marlin!

Comment (qharley) [inline src/modules/tools/zprobe/ZGridStrategy.cpp]: Done

File: src/modules/tools/zprobe/ZGridStrategy.h [hunk 4]
```diff
@@ -45,6 +45,8 @@ class ZGridStrategy : public LevelingStrategy
 
     bool wait_for_probe;
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/ZGridStrategy.h]: Should go in the struct for bools to save RAM.

File: src/modules/tools/zprobe/ZGridStrategy.cpp [hunk 5]
```diff
@@ -44,6 +44,10 @@
 
        leveling-strategy.ZGrid-leveling.wait_for_probe  true
 
+    The machine can be told to home in one of the following modes:
+
+       leveling-strategy.ZGrid-leveling.home_before_probe  2;    # 0: NOHOME  1: HOMEXY  2: HOMEXYZ (Default HOMEXYZ)
```
Comment (wolfmanjm) [inline src/modules/tools/zprobe/ZGridStrategy.cpp]: I'm nitpicking here but why use 0, 1, 2 and not nohome, homexy, hmexyz in the config, it is much clearer, and I prefer that to cryptic numbers in the config.

Comment (qharley) [inline src/modules/tools/zprobe/ZGridStrategy.cpp]: fixing that up right now

**Comment by qharley** [comment]
should be ok now?

**Comment by wolfmanjm** [comment]
I cleaned it up a little, it is easier to do that myself than comment on it (although I already commented on one of them, which you missed ;)

**Comment by qharley** [comment]
No problem.  After all, I am not really a programmer...

Thanks,
  [image: photo]
*Quentin Harley*
Founder, Reprap Morgan 3D
 p:+27120041883 | e:quentin@reprapmorgan.com | w:www.reprapmorgan.com | a:4
Burger Ave, Lyttelton Manor, Centurion

**Pull request merged**